File: frontend/localtrade_app/lib/core/models/product_model.dart
class Product {
  final String id;
  final String title;
  final String description;
  final String category;
  final double price;
  final List<String> images;
  final String vendorId;
  final String? vendorName;
  final String? location;
  final int stockQuantity;
  final String productStatus;
  final double ratingsAverage;
  final int ratingsQuantity;
  final DateTime? createdAt;

  Product({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.price,
    required this.images,
    required this.vendorId,
    this.vendorName,
    this.location,
    this.stockQuantity = 0,
    this.productStatus = 'Available',
    this.ratingsAverage = 0,
    this.ratingsQuantity = 0,
    this.createdAt,
  });

  bool get isAvailable => productStatus == 'Available' && stockQuantity > 0;
  bool get isOutOfStock => productStatus == 'OutOfStock' || stockQuantity <= 0;
  bool get isInactive => productStatus == 'Inactive';

  String get displayPrice => 'Rs. ${price.toStringAsFixed(price % 1 == 0 ? 0 : 2)}';
  String get stockLabel => isOutOfStock ? 'Out of Stock' : '$stockQuantity in stock';
  String get categoryEmoji {
    switch (category) {
      case 'Vegetables': return '🥬';
      case 'Dairy': return '🥛';
      case 'Handicrafts': return '🎨';
      case 'Clothing': return '👕';
      case 'Local Goods': return '🌾';
      case 'Tailoring': return '✂️';
      case 'Groceries': return '🧺';
      case 'Bakery': return '🍞';
      case 'Meat': return '🥩';
      default: return '📦';
    }
  }

  String get primaryImage => images.isNotEmpty ? images.first : '';

  factory Product.fromJson(Map<String, dynamic> json) {
    final vendorIdRaw = json['vendorId'];
    String vendorId;
    String? vendorName;

    if (vendorIdRaw is Map) {
      vendorId = vendorIdRaw['_id'] ?? '';
      vendorName = vendorIdRaw['shopName'] ?? vendorIdRaw['fullName'];
    } else {
      vendorId = vendorIdRaw ?? '';
      vendorName = json['vendorName'];
    }

    return Product(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? 'Others',
      price: (json['price'] ?? 0).toDouble(),
      images: List<String>.from(json['images'] ?? []),
      vendorId: vendorId,
      vendorName: vendorName,
      location: json['location'],
      stockQuantity: json['stockQuantity'] ?? json['stock'] ?? 0,
      productStatus: json['productStatus'] ?? 'Available',
      ratingsAverage: (json['ratingsAverage'] ?? 0).toDouble(),
      ratingsQuantity: json['ratingsQuantity'] ?? 0,
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'category': category,
      'price': price,
      'images': images,
      'vendorId': vendorId,
      'vendorName': vendorName,
      'location': location,
      'stockQuantity': stockQuantity,
      'productStatus': productStatus,
      'ratingsAverage': ratingsAverage,
      'ratingsQuantity': ratingsQuantity,
    };
  }
}

File: backend/src/models/productModel.js
const mongoose = require('mongoose');

const productSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: [true, 'A product must have a title'],
      trim: true,
      maxlength: [100, 'A product title must have less or equal than 100 characters'],
    },
    description: {
      type: String,
      required: [true, 'A product must have a description'],
      trim: true,
    },
    category: {
      type: String,
      required: [true, 'A product must have a category'],
      trim: true,
    },
    price: {
      type: Number,
      required: [true, 'A product must have a price'],
    },
    originalPrice: {
      type: Number,
      default: null,
    },
    priceUnit: {
      type: String,
      enum: ['piece', 'kg', '100g', 'liter', 'dozen', 'packet', 'bundle'],
      default: 'piece',
    },
    minOrder: {
      type: Number,
      default: 1,
      min: [0.1, 'Minimum order must be at least 0.1'],
    },
    images: {
      type: [String],
      required: [true, 'A product must have at least one image'],
    },
    vendorId: {
      type: mongoose.Schema.ObjectId,
      ref: 'User',
      required: [true, 'Product must belong to a vendor'],
    },
    stockQuantity: {
      type: Number,
      required: [true, 'Please provide stock quantity'],
      min: [0, 'Stock cannot be negative'],
      default: 0,
    },
    productStatus: {
      type: String,
      enum: ['Available', 'OutOfStock', 'Inactive'],
      default: 'Available',
    },
    vendorName: {
      type: String,
      trim: true,
    },
    location: {
      type: String,
      trim: true,
    },
    ratingsAverage: {
      type: Number,
      default: 0,
      min: [0, 'Rating must be above or equal to 0'],
      max: [5, 'Rating must be below or equal to 5'],
      set: val => Math.round(val * 10) / 10, // rounds to 4.7
    },
    ratingsQuantity: {
      type: Number,
      default: 0,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Virtuals for backward compatibility
productSchema.virtual('stock').get(function() {
  return this.stockQuantity;
}).set(function(val) {
  this.stockQuantity = val;
});

productSchema.virtual('isAvailable').get(function() {
  return this.productStatus === 'Available' && this.stockQuantity > 0;
});

// Middleware to update status based on stock — only if productStatus wasn't explicitly set
productSchema.pre('save', function() {
  if (this.isModified('productStatus')) return;
  if (this.stockQuantity <= 0) {
    this.productStatus = 'OutOfStock';
  } else if (this.productStatus === 'OutOfStock') {
    this.productStatus = 'Available';
  }
});

// Indexes for search and filter performance
productSchema.index({ title: 'text', description: 'text', vendorName: 'text' });
productSchema.index({ category: 1 });
productSchema.index({ vendorId: 1 });
productSchema.index({ price: 1 });
productSchema.index({ productStatus: 1 });
productSchema.index({ stockQuantity: 1 });
productSchema.index({ location: 1 });
productSchema.index({ createdAt: -1 });

const Product = mongoose.model('Product', productSchema);

module.exports = Product;
